File: MPC/implementation/mpc.py
import numpy as np
import matplotlib.pyplot as plt
import cvxpy as cp

# Define the system dynamics (discrete-time linear system)
dt = 0.1  # Time step
A = np.array([[1, dt], [0, 1]])  # State transition matrix
B = np.array([[0], [dt]])  # Control input matrix
n = A.shape[0]  # Number of states
m = B.shape[1]  # Number of control inputs

# MPC parameters
N = 10  # Prediction horizon
x_goal = np.array([10, 0])  # Target state

# Define the cost matrices
Q = np.eye(n)  # State cost matrix
R = 0.1 * np.eye(m)  # Control input cost matrix

# Constraints
u_min, u_max = -2, 2
x_min, x_max = -np.inf, np.inf

# Initialize state and control trajectory
x = np.array([0, 0])  # Initial state
u_traj = []

# Simulation
T = 50  # Total simulation time steps
x_traj = [x]

for _ in range(T):
    # Define optimization variables
    x_var = cp.Variable((n, N + 1))
    u_var = cp.Variable((m, N))

    # Define the cost function
    cost = 0
    constraints = []
    for t in range(N):
        cost += cp.quad_form(x_var[:, t] - x_goal, Q) + cp.quad_form(u_var[:, t], R)
        constraints += [x_var[:, t + 1] == A @ x_var[:, t] + B @ u_var[:, t]]
        constraints += [u_min <= u_var[:, t], u_var[:, t] <= u_max]

    # Terminal cost
    cost += cp.quad_form(x_var[:, N] - x_goal, Q)

    # Initial state constraint
    constraints += [x_var[:, 0] == x]

    # State constraints
    constraints += [x_min <= x_var, x_var <= x_max]

    # Solve the optimization problem
    problem = cp.Problem(cp.Minimize(cost), constraints)
    problem.solve()

    # Apply the first control input
    u_opt = u_var[:, 0].value
    x = A @ x + B @ u_opt
    x_traj.append(x)
    u_traj.append(u_opt)

# Convert trajectories to arrays for plotting
x_traj = np.array(x_traj)
u_traj = np.array(u_traj)

# Plot results
plt.figure(figsize=(12, 5))

plt.subplot(1, 2, 1)
plt.plot(x_traj[:, 0], label='Position')
plt.plot([x_goal[0]] * len(x_traj), 'r--', label='Target Position')
plt.xlabel('Time step')
plt.ylabel('Position')
plt.title('State Trajectory')
plt.legend()

plt.subplot(1, 2, 2)
plt.step(range(len(u_traj)), u_traj, where='post')
plt.xlabel('Time step')
plt.ylabel('Control Input')
plt.title('Control Input')

plt.tight_layout()
plt.show()

File: MPC/implementation/mpc.cpp
#include <Eigen/Dense>
#include <qpOASES.hpp>
#include <vector>
#include <iostream>
#include <cmath>
#include <iomanip>

using namespace Eigen;
using namespace qpOASES;

int main() {
    const double dt = 0.1;  // Time step

    // System dynamics
    Matrix2d A;
    A << 1, dt, 0, 1;

    Vector2d B;
    B << 0, dt;

    const int n = 2;  // Number of states
    const int m = 1;  // Number of control inputs

    // MPC parameters
    const int N = 10;  // Prediction horizon
    Vector2d x_goal;
    x_goal << 10, 0;  // Target state

    // Cost matrices
    Matrix2d Q = Matrix2d::Identity();  // State cost matrix
    MatrixXd R = 0.1 * MatrixXd::Identity(m, m);  // Control input cost matrix

    // Constraints
    const double u_min = -2, u_max = 2;

    // Initial state
    Vector2d x;
    x << 0, 0;

    // Store state and control trajectories
    std::vector<Vector2d> x_traj;
    std::vector<double> u_traj;
    x_traj.push_back(x);

    // Total simulation time steps
    const int T = 50;

    for (int t = 0; t < T; ++t) {
        // Set up QP problem
        MatrixXd H = MatrixXd::Zero(N * (n + m), N * (n + m));
        VectorXd g = VectorXd::Zero(N * (n + m));

        // Fill cost matrix and vector
        for (int i = 0; i < N; ++i) {
            H.block(i * n, i * n, n, n) = Q;  // State cost
            H.block(N * n + i * m, N * n + i * m, m, m) = R;  // Input cost
            g.segment(i * n, n) = -2.0 * Q * x_goal;  // Linear cost term for state
        }
        g.segment(N * n, m) = VectorXd::Zero(m);

        // Fill equality constraints (dynamics)
        MatrixXd Aeq = MatrixXd::Zero(N * n, N * (n + m));
        VectorXd beq = VectorXd::Zero(N * n);

        for (int i = 0; i < N; ++i) {
            if (i == 0) {
                Aeq.block(0, 0, n, n) = MatrixXd::Identity(n, n);
                beq.segment(0, n) = x;
            } else {
                Aeq.block(i * n, (i - 1) * n, n, n) = -A;
                Aeq.block(i * n, N * n + (i - 1) * m, n, m) = -B;
                Aeq.block(i * n, i * n, n, n) = MatrixXd::Identity(n, n);
            }
        }

        // Input constraints
        MatrixXd Ain = MatrixXd::Zero(N * m, N * (n + m));
        VectorXd lb = VectorXd::Constant(N * m, u_min);
        VectorXd ub = VectorXd::Constant(N * m, u_max);

        for (int i = 0; i < N; ++i) {
            Ain.block(i * m, N * n + i * m, m, m) = MatrixXd::Identity(m, m);
        }

        // Solve QP
        QProblem qp(N * (n + m), N * m);
        Options options;
        options.setToDefault();
        qp.setOptions(options);

        int nWSR = 100;
        qp.init(H.data(), g.data(), Ain.data(), lb.data(), ub.data(), Aeq.data(), beq.data(), nWSR);

        VectorXd z(N * (n + m));
        qp.getPrimalSolution(z.data());

        // Extract control input
        double u_opt = z[N * n];

        // Apply control input to the system
        x = A * x + B * u_opt;
        x_traj.push_back(x);
        u_traj.push_back(u_opt);
    }

    // Print results
    std::cout << "State Trajectory:\n";
    for (const auto& state : x_traj) {
        std::cout << state.transpose() << std::endl;
    }

    std::cout << "Control Trajectory:\n";
    for (const auto& u : u_traj) {
        std::cout << u << std::endl;
    }

    return 0;
}
